sdk/core/Azure.Core/src/Shared/TaskExtensions.cs

```diff
@@ -87,5 +91,39 @@ private static void VerifyTaskCompleted(bool isCompleted)
                 throw new InvalidOperationException("Task is not completed");
             }
         }
+
+        public readonly struct Enumerable<T> : IEnumerable<T>
+        {
+            private readonly IAsyncEnumerable<T> _asyncEnumerable;
+
+            public Enumerable(IAsyncEnumerable<T> asyncEnumerable) => _asyncEnumerable = asyncEnumerable;
+
+            public Enumerator<T> GetEnumerator() => new Enumerator<T>(_asyncEnumerable.GetAsyncEnumerator());
+
+            IEnumerator<T> IEnumerable<T>.GetEnumerator() => new Enumerator<T>(_asyncEnumerable.GetAsyncEnumerator());
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+        }
+
+        public readonly struct Enumerator<T> : IEnumerator<T>
+        {
+            private readonly IAsyncEnumerator<T> _asyncEnumerator;
+
+            public Enumerator(IAsyncEnumerator<T> asyncEnumerator) => _asyncEnumerator = asyncEnumerator;
+
+#pragma warning disable AZC0107 // DO NOT call public asynchronous method in synchronous scope.
+            public bool MoveNext() => _asyncEnumerator.MoveNextAsync().EnsureCompleted();
```

Comment:
I may be misreading, but wouldn't this potentially cause issues in DEBUG builds because the task could be uncompleted?   Do we need an inner implementation that calls `Wait` or similar before using `EnsureCompleted`?

Reply:
It looks like you've added a few methods for this case, but aren't using them in these types.   Was that intentional?

Reply:
> I may be misreading, but wouldn't this potentially cause issues in DEBUG builds because the task could be uncompleted?

Then we have an issue and it must be investigated. `EnsureSyncEnumerable` works similarly to `EnsureCompleted`, it ensures that enumerable is actually synchronous. We can't check it immediately for the whole enumerable, we have to check it every time when `MoveNext()` is called (`MoveNextAsync` of the underlying `IAsyncEnumerator<T>` should return completed task).

Reply:
> It looks like you've added a few methods for this case, but aren't using them in these types.

I'm not sure I understand what you mean. Which methods exactly?

Reply:
I was referring to `EnsureCompleted` that takes a `ConfiguredValueTaskAwaitable` and `async` and the `EnsureSyncEnumerable<T>` methods,  though from discussion it would seem the intent for the latter of those is for library authors to call directly and not this plumbing.

Reply:
- `EnsureSyncEnumerable` is used in `PartitionedUploader` and `DataLakePartitionedUploader`
- `EnsureCompleted` with `ConfiguredValueTaskAwaitable` parameter wasn't added in this PR